File: scarl-ui/src/system/Spaceship.js
import { Record } from 'immutable'
import * as coords from './coordinates'
import { hypotenuse, vector } from './physics'
import Position from './Position'
import Vector from './Vector'

const Travel = Record({
    destination: undefined,
    from: Position(),
    to: Position(),
    position: Position(),
    speed: Vector({x: 0, y: 0}),
    eta: undefined,
})
Travel.read = ({from, to, position, speed, ...data}) => Spaceship({
    ...data,
    from: Position.read(from),
    to: Position.read(to),
    position: Position.read(position),
    speed: Vector.read(speed),
})
Travel.distance = travel => hypotenuse(travel.to.x - travel.from.x, travel.to.y - travel.from.y)
Travel.finished = travel => {
    const quarter = coords.quarter(travel.from, travel.to)

    const p1 = coords.normalize(quarter, travel.position)
    const p2 = coords.normalize(quarter, travel.to)

    const dx = p2.x - p1.x

    // destination reached or speed in reverse
    return dx <= 0 || coords.normalize(quarter, travel.speed).x < 0
}

const Spaceship = Record({
    id: undefined,
    source: undefined,
    acceleration: undefined,
    travel: null,
    port: null,
})
Spaceship.read = ({travel, ...data}) => Spaceship({
    ...data,
    travel: travel ? Travel.read(travel) : null,
})
Spaceship.tick = tick => ship => {
    if (! ship.travel) {
        return ship
    }
    const acceleration = calculateTravelAcceleration(ship)

    const speed = Vector.add(ship.travel.speed, Vector.multiply(acceleration, tick))
    const position = Position.add(ship.travel.position, Vector.multiply(speed, tick))

    const next = ship.travel.set('position', position).set('speed', speed)

    return Travel.finished(next) ? (
        ship.set('travel', null).set('port', next.destination)
    ) : (
        ship.set('travel', next).set('port', null)
    )
}

const calculateTravelAcceleration = ship => {
    const quarter = coords.quarter(ship.travel.from, ship.travel.to)
    const p1 = coords.normalize(quarter, ship.travel.position)
    const p2 = coords.normalize(quarter, ship.travel.to)

    const dx = p2.x - p1.x
    const dy = p2.y - p1.y

    const distance = hypotenuse(dx, dy)
    const half = Travel.distance(ship.travel) / 2
    const a = dx < 0 || distance < half ? -ship.acceleration : ship.acceleration

    return coords.normalize(quarter, vector(dy/dx, a))
}

export { Travel }
export default Spaceship

File: scarl/src/main/scala/io/github/fiifoo/scarl/world/system/Spaceship.scala
package io.github.fiifoo.scarl.world.system

import io.github.fiifoo.scarl.world.system.Spaceship.Travel
import io.github.fiifoo.scarl.world.system.source.SpaceshipSourceId

object Spaceship {
  type Id = String

  case class Travel(destination: StellarBody.Id,
                    from: Position,
                    to: Position,
                    position: Position,
                    speed: Vector = Vector(0, 0),
                   ) {
    def distance: Double = Physics.hypotenuse(this.to.x - this.from.x, this.to.y - this.from.y)

    def finished: Boolean = {
      val quarter = Coordinates.Quarter(this.from, this.to)

      // speed stopped or in reverse
      quarter.normalize(this.speed).x <= 0
    }
  }

}

case class Spaceship(id: Spaceship.Id,
                     source: SpaceshipSourceId,
                     acceleration: Double,
                     travel: Option[Travel] = None,
                     port: Option[StellarBody.Id] = None,
                    ) {

  def tick(tick: Int): Spaceship = {
    this.travel map calculateTravel(tick) getOrElse this
  }

  private def calculateTravel(tick: Int)(travel: Travel): Spaceship = {
    val acceleration = this.calculateTravelAcceleration(travel)

    val speed = travel.speed.add(acceleration.multiply(tick))
    val position = travel.position.add(speed.multiply(tick))

    val next = travel.copy(
      position = position,
      speed = speed
    )

    if (next.finished) {
      this.copy(
        travel = None,
        port = Some(travel.destination)
      )
    } else {
      this.copy(
        travel = Some(next),
        port = None
      )
    }
  }

  private def calculateTravelAcceleration(travel: Travel): Vector = {
    val quarter = Coordinates.Quarter(travel.from, travel.to)
    val p1 = quarter.normalize(travel.position)
    val p2 = quarter.normalize(travel.to)

    val dx = p2.x - p1.x
    val dy = p2.y - p1.y

    val distance = Physics.hypotenuse(dx, dy)
    val half = travel.distance / 2
    val a = if (dx < 0 || distance < half) -this.acceleration else this.acceleration

    quarter.normalize(Physics.vector(dy / dx, a))
  }
}
